pkg/scheduler/podtask/ranges.go

```diff
@@ -0,0 +1,140 @@
+package podtask
+
+import (
+	"sort"
+
+	"github.com/gogo/protobuf/proto"
+	mesos "github.com/mesos/mesos-go/mesosproto"
+)
+
+// Range is an all-inclusive range defined by a lower and upper bound.
+type Range [2]uint64
+
+// Ranges represents a list of Ranges.
+type Ranges []Range
+
+// NewPortRanges returns Ranges from the "ports" resource in the
+// given *mesos.Offer. If that resource isn't provided, nil will be returned.
+//
+// The returned Ranges are sorted and have all overlapping ranges merged from
+// left to right. e.g. [[0, 5], [4, 3], [10, 7]] -> [[0, 5], [7, 10]]
+func NewPortRanges(o *mesos.Offer) Ranges {
+	if o == nil {
+		return nil
+	}
+
+	var r *mesos.Resource
+	for i := range o.Resources {
+		if o.Resources[i].GetName() == "ports" {
+			r = o.Resources[i]
+			break
+		}
+	}
+
+	if r == nil {
+		return nil
+	}
+
+	offered := r.GetRanges().GetRange()
+	rs := make(Ranges, len(offered))
+	for i, r := range offered {
+		if lo, hi := r.GetBegin(), r.GetEnd(); lo <= hi {
+			rs[i][0], rs[i][1] = lo, hi
+		} else {
+			rs[i][0], rs[i][1] = hi, lo
+		}
+	}
+	sort.Sort(rs)
+
+	return rs.Squash()
+}
+
+// These three methods implement sort.Interface
+func (rs Ranges) Len() int           { return len(rs) }
+func (rs Ranges) Less(i, j int) bool { return rs[i][0] < rs[j][0] && rs[i][1] < rs[j][1] }
+func (rs Ranges) Swap(i, j int)      { rs[i], rs[j] = rs[j], rs[i] }
+
+// Size returns the sum of the Size of all Ranges.
+func (rs Ranges) Size() uint64 {
+	var sz uint64
+	for i := range rs {
+		sz += 1 + (rs[i][1] - rs[i][0])
+	}
+	return sz
+}
+
+// Squash merges overlapping and continuous Ranges. It assumes they're pre-sorted.
+func (rs Ranges) Squash() Ranges {
+	if len(rs) < 2 {
+		return rs
+	}
+	squashed := Ranges{rs[0]}
+	for i := 1; i < len(rs); i++ {
+		switch top := squashed[len(squashed)-1]; {
+		case 1+top[1] < rs[i][0]: // no overlap nor continuity: push
+			squashed = append(squashed, rs[i])
+		case 1+top[1] <= rs[i][1]: // overlap or continuity: squash
+			squashed[len(squashed)-1][1] = rs[i][1]
+		}
+	}
+	return squashed
+}
+
+// Find performs a binary search for n returning the index of the Range it was
+// found at or -1 if not found.
+func (rs Ranges) Find(n uint64) int {
```

Comment:
I misread `Find` at first because I thought it was finding a port, not a port range. Could we rename to `RangeFor` for `RangeIndexOf`, or something to indicate that it's returning a range index?